### PR TITLE
New level of abstraction for streams with (non-canonical) location function

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** cleanup

The PR introduces a new module `LStream` to abstract over the imperative details of manipulating a stream equipped with a non-canonical location function, such as the one associated to the lexer.

This allows to save some side-effects in `CLexer` (the hash table of locations) and in `Grammar` (the global references `floc` and `token_count`). This allows also to hide the `Plexing.location_function` which some stream-related functions (but not all) were carrying.

We clarify/simplify a bit the error reporting in `Grammar.Entry.parsable`.

Some arbitrary choices are made, e.g. the name `LStream` itself and its location in `lib` (since it depends on `loc.ml`), etc... Is it ok?

- [X] Corresponding ocaml documentation was added / updated

### CODE REVIEW

**Comment by ppedrot** [comment]
I like this, it makes the code simpler. @herbelin You need to fix the file dependency though, so this stream library might need to live somewhere else.

**Comment by herbelin** [comment]
I made a change (actually recording max visited part of the stream, as intented) which should fix #14098 (locating longest visited part of stream in case of failure in the middle of a look-ahead).

**Comment by ppedrot** [comment]
@herbelin is elpi broken for a real reason? Something fishy seems to be happening, it reliably fails with a SIGKILL. Maybe you could try to rebase on master to see if it persists?

**Comment by herbelin** [comment]
Elpi needed an overlay. Unfortunately, the change in the type of `Entry.of_parser` (from `string -> (Plexing.location_function -> te LStream.t -> 'a) -> 'a t` to `string -> (te LStream.t -> 'a) -> 'a t`  is not detected by OCaml because of the polymophic `'a`. So, this had to be checked manually. I added a note in `dev/doc/changes.md` about it.

**Comment by ppedrot** [comment]
@herbelin while you're changing this area of the code it might be worthwhile to wrap the closure argument of `of_entry` into a record to prevent future instances of this kind of issue.

**Comment by herbelin** [comment]
That's a very good idea.

Do you mean something like that:
```
    type 'a parser_fun = { parser_fun : te LStream.t -> 'a }                                                             
    val of_parser : string -> 'a parser_fun -> 'a t                                                             
```
so that compilation necessarily breaks.

**Comment by ppedrot** [comment]
@herbelin indeed.

File: dev/doc/changes.md [hunk 1]
```diff
@@ -1,3 +1,11 @@
+## Changes between Coq 8.13 and Coq 8.14
+
+Gramlib
+
+- A few functions change their interfaces to take benefit of a new abstraction level `LStream` for streams with location function.
+
+  Warning: the type of `Pcoq.Entry.of_parser` changed from `string -> (Plexing.location_function -> te LStream.t -> 'a) -> 'a t` to `string -> (te LStream.t -> 'a) -> 'a t` but OCaml's type-checker generally do not detect it. Existing files compile but then typically raise a stack overflow when parsing. The changes to calls to `Pcoq.Entry.of_parser` have then to be done by "manual" scrutinization.
```
Comment (ppedrot) [inline dev/doc/changes.md:7]: *the* OCaml type-checker generally *does not*

Comment (ppedrot) [reply dev/doc/changes.md:7]: Actually this comment is outdated, so you can probably remove it altogether.

Comment (herbelin) [reply dev/doc/changes.md]: I removed the outdated comment.

**Review by ppedrot — APPROVED** [review]

**Pull request merged**